evaluation/numpy/segmentation_and_tracking_quality.py

```diff
@@ -0,0 +1,290 @@
+          semantic_label != self._ignore_label, semantic_label,
+          self._num_classes)
+      semantic_prediction = np.where(
+          semantic_prediction != self._ignore_label,
+          semantic_prediction, self._num_classes)
+    if sequence_id in self._iou_confusion_matrix_per_sequence:
+      idxs = (np.reshape(semantic_label, [-1]) << self._label_bit_shift) + np.reshape(
+          semantic_prediction, [-1])
+      unique_idxs, counts = np.unique(idxs, return_counts=True)
+      self._iou_confusion_matrix_per_sequence[sequence_id][
+          unique_idxs >> self._label_bit_shift, unique_idxs & self._bit_mask] += counts
+      self._sequence_length[sequence_id] += 1
+    else:
+      self._iou_confusion_matrix_per_sequence[sequence_id] = np.zeros(
+          (self._confusion_matrix_size, self._confusion_matrix_size), dtype=np.int64)
+      idxs = np.stack([np.reshape(semantic_label, [-1]), 
+                np.reshape(semantic_prediction, [-1])], axis=0)
+      np.add.at(self._iou_confusion_matrix_per_sequence[sequence_id], tuple(idxs), 1)
+
+      self._predictions[sequence_id] = {}
+      self._ground_truth[sequence_id] = {}
+      self._intersections[sequence_id] = {}
+      self._sequence_length[sequence_id] = 1
+
+    # instance_label = y_true % self._max_instances_per_category
+    instance_label = y_true & self._bit_mask # 0xFFFF == 2 ^ 16 - 1
+
+    label_mask = np.zeros_like(semantic_label, dtype=np.bool)
+    prediction_mask = np.zeros_like(semantic_prediction, dtype=np.bool)
+    for things_class_id in self._things_list:
+      label_mask = np.logical_or(label_mask,
+                                 semantic_label == things_class_id)
+      prediction_mask = np.logical_or(
+          prediction_mask, semantic_prediction == things_class_id)
+
+    # Select the `crowd` region of the current class. This region is encoded
+    # instance id `0`.
+    is_crowd = np.logical_and(instance_label == 0, label_mask)
+    # Select the non-crowd region of the corresponding class as the `crowd`
+    # region is ignored for the tracking term.
+    label_mask = np.logical_and(label_mask, np.logical_not(is_crowd))
+    # Do not punish id assignment for regions that are annotated as `crowd` in
+    # the ground-truth.
+    prediction_mask = np.logical_and(prediction_mask, np.logical_not(is_crowd))
+
+    seq_preds = self._predictions[sequence_id]
+    seq_gts = self._ground_truth[sequence_id]
+    seq_intersects = self._intersections[sequence_id]
+
+    # Compute and update areas of ground-truth, predictions and intersections.
+    _update_dict_stats(seq_preds, y_pred[prediction_mask])
+    _update_dict_stats(seq_gts, y_true[label_mask])
+
+    non_crowd_intersection = np.logical_and(label_mask, prediction_mask)
+    intersection_ids = (
+        y_true[non_crowd_intersection] * self._offset +
+        y_pred[non_crowd_intersection])
+    _update_dict_stats(seq_intersects, intersection_ids)
+
+  def result(self) -> Dict[Text, Any]:
```

Comment:
Mapping[...]? Use immutable if possible

Reply:
Switched to mapping. I am not sure which immutable dict-like type you're referring to. I tried to recreate the TF interface as close as possible.

Reply:
Yes Mapping should be good.